### ISSUE TITLE
Inconsistencies due to when clients are created

### ISSUE THREAD

**Issue by jakearchibald** [w3c/ServiceWorker#870]
https://jakearchibald.github.io/isserviceworkerready/demos/clients-count/
1. Visit the above
2. Hit shift + refresh
3. Hit refresh

Chrome says 1, Firefox says 0.

The Firefox behaviour seems right to me, but unless I'm reading it wrong, the spec isn't clear around this. This might be because of earlier confusion around when clients are created.

From https://html.spec.whatwg.org/multipage/browsers.html#navigating-across-documents:

> Set request's client to the source browsing context's active document's Window object's environment settings object

If I'm reading this correctly, the client is the window being navigated, not a new client ready for the new page.

From https://slightlyoff.github.io/ServiceWorker/spec/service_worker/#on-fetch-request-algorithm 12.1:

> If client is not a secure context, return null.

This seems to assume that the client is a new client ready for the new page, but doesn't this mean that SW will fail if you're navigating _from_ an unrelated insecure page?

**Comment by jakearchibald** [comment]
Related Chrome bug https://bugs.chromium.org/p/chromium/issues/detail?id=514148

**Comment by wanderview** [comment]
I think this is a Firefox bug and chrome is more correct.

The current spec says to control windows.  And it says to associate the service worker with the window (i.e. control it)  before firing the FetchEvent.  So there should be an existing controlled client when a navigation FetchEvent is fired.

The problem in Firefox is that we control documents, not windows.  The document does not exist before the navigation, so we don't have a controlling client yet.  There is a bug open for this, but there was some discussion at the last f2f that we might want to change to controlling documents in the spec.  So we have not moved on this.

  https://bugzilla.mozilla.org/show_bug.cgi?id=1237498

In regards to this spec language:

> If client is not a secure context, return null.

I think the spec is trying to check for a non-secure parent window.  It probably needs to be re-worded since it should use the navigation URL security and parent window security to make this determination.

**Comment by jakearchibald** [comment]
We should talk about this at the f2f. I think the Firefox behaviour makes more sense, especially given what we do with `fetchEvent.clientId`.

**Comment by wanderview** [comment]
If we decide to keep the "client is Window" behavior currently in spec, then I think we could revisit providing a value in `fetchEvent.clientId` for navigations.  My objections there were largely based on my misconception about clients being documents.

**Comment by jakearchibald** [comment]
This issue is very related to https://github.com/slightlyoff/ServiceWorker/issues/811

**Comment by jakearchibald** [comment]
F2F: 

* A reserved client is created for a navigation
* `fetchEvent.clientId` - represents the client that initiated the request
* `fetchEvent.reservedClientId` - represents the client that has been created for potentially-client-creating request, which may not be used in the event of a redirect, content-disposition, network error, or other response rejection
* `fetchEvent.targetClientId` - the id of the client that the new client will replace if it's used
* `clients.matchAll()` will not return reserved clients
* `clients.matchAll({ includeReserved: true })` will return reserved clients
* Reserved clients will have url `about:blank`
* Reserved clients will inherit visual properties from their target client (which may be a cross origin client, or new tab)
* `client.used` will be false for reserved clients

Some of the naming is up for bikeshedding, specifically:

* "reserved", in all the properties it's used
* `targetClientId`
* `client.used`

**Comment by jakearchibald** [comment]
`openedWindow.location.href = 'blah'` - what client initiates this request? The `openedWindow` or the window which executed the code?

**Comment by jakearchibald** [comment]
`targetClientId` will be the same as `reservedClientId` if the link is being opened in a new window.

**Comment by wanderview** [comment]
@jakearchibald Can we mark this one decided?  I thought we decided it in the meeting.

**Comment by wanderview** [comment]
Gecko bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1264177

**Comment by mattto** [comment]
Chromium bug: https://bugs.chromium.org/p/chromium/issues/detail?id=514148

**Comment by mkruisselbrink** [comment]
Since we didn't explicitly discuss this, I wonder if what we came up with also sufficiently addresses worker clients. For workers their settings object (client?) is created before the worker is fetched, and that worker is then used as the client for the actual request to fetch the worker.

So would that be an exception to the `fetchEvent.clientId` represents the client that initiated the request rule? Or since it's not a navigation (but it is a new controlled client) would we only have a `clientId` and have both `reservedClientId` and `targetClientId` be null/undefined?

**Comment by jungkees** [comment]
> For workers their settings object (client?) is created before the worker is fetched, and that worker is then used as the client for the actual request to fetch the worker.

I think this is not true. The fetch in run a worker is invoked with _outside settings_ which is the source client.

**Comment by jungkees** [comment]
I outlined the redefinition of the service worker client concept: https://gist.github.com/jungkees/03d4473ae47bd222539ab9cf5c9afd31

Hooks to HTML's navigation and run a worker will be needed (with some changes to https://github.com/whatwg/html/pull/322).

In the gist proposal, I suggested we have a single .potentialClientId (with its state defined) over both .reservedClientId and .targetClientId.

Please take a look.

/cc @annevk @jakearchibald @slightlyoff @mkruisselbrink @wanderview @ehsan

**Comment by annevk** [comment]
I don't really understand the motivation. It says a service worker can outlive a client, and then it goes on about how a service worker exists before a client is created.

How does this work in browsers today?

Supposedly we always know when we are creating a new client, so we could generate an ID that we then later use once the client is created. Is that what you're trying to accomplish here?

**Comment by jungkees** [comment]
> so we could generate an ID that we then later use once the client is created.

This is what I suggested.

1. Navigation is triggered.
1. [Navigation] Invoke fetch.
1. [Fetch] Invoke Handle Fetch.
1. [Handle Fetch] Create (potential) target client information
1. [Navigation] Create (target) Window/Document
1. [Navigation] Set Window/Document's client information to the value created in 4.

In order to do that, the step 4 and 6 above should be added. And for this, I wanted to have some internal slots for id and active worker on a fetch algorithm instance or on a request.

> How does this work in browsers today?

Blink creates some corresponding state (object) when fetch occurs, and the properties in that object are later associated with the the page/worker. So, largely the steps I suggest here is what the implementation does now.

**Comment by annevk** [comment]
I think it would be better if we created the ID between step 1 and 2 and pass that along in the request. After all, navigate ends up creating the document which is presumably where this information is stored eventually.

**Comment by annevk** [comment]
(And a similar setup for workers.)

**Comment by jungkees** [comment]
> it would be better if we created the ID between step 1 and 2 and pass that along in the request.

This sounds fine to me.

How about the matched active worker? Handle fetch should be the best place for this I think. Any better idea?

**Comment by annevk** [comment]
That would mean it would have to somehow make its way back up the chain (e.g., by sticking on a response), which seems a little ugly. What is needed to determine the active worker?

**Comment by jungkees** [comment]
Right. Determining the active worker is done in [Handle Fetch](https://slightlyoff.github.io/ServiceWorker/spec/service_worker/index.html#on-fetch-request-algorithm) step 12.3. That is, the fetch is a non-subresource request and the scope of the request url matches one of the registrations.

**Comment by annevk** [comment]
Okay, so the worry is about some kind of race condition if we match twice on the URL? I think you have mentioned that earlier.

The operations performed on client there don't make sense, right? Client doesn't exist yet for non-subresource requests.

I think here too it makes more sense if we end up dictating what service worker to use. Rather than letting "Handle Fetch" figure it out somehow and tell us about it later what it came up. 

It also seems plausible we might have more such cases in the future where you'd want to fetch something using a particular service worker.

One way to do this would be to change the "skip-service-worker flag" of request into a "service worker" setting. Which is one of "none", "client", or a service worker instance of sorts.

**Comment by wanderview** [comment]
Note, we're unlikely to get any review here from google folks while google I/O is going on this week.

**Comment by jungkees** [comment]
@annevk I'm looking at more of how the implementations currently work. Blink finds matched registration in navigation algorithm equivalent phase and get its active worker right before dispatching the fetch event (in handle fetch algorithm equivalent phase). Your suggestion seems okay to me now. Let me try to examine a bit more next week. @wanderview Google folks will be able to have a look by then. On the spec side, Anne and I can discuss, and on the API side, more discussion would be needed I guess.

**Comment by jungkees** [comment]
I think we should rather set the matched service worker **registration** to a request rather than the active worker. The active worker can change (by waiting worker being promoted) between the step in navigation and the step later in Handle Fetch if the registration has had both the active worker and the waiting worker.

Also, in the same reason, the active worker that will be attached to the client (document or other worker) should be determined when the registration's active worker for the fetch event is determined (in Handle Fetch). So, it seems to me that the active worker should be passed back to the caller within a response. Or can we create some outer object in navigate that has the active worker slot and pass the reference to fetch/Handle Fetch and change its state in the Handle Fetch?

**Comment by annevk** [comment]
That seems like an unfortunate setup since in general fetch doesn't need to know about this, only for navigation.

**Comment by jakearchibald** [comment]
Pre F2F notes: Does the new proposal let us distinguish between:

* A same-tab navigation
* An open-in-new-window navigation
* An open-in-another-existing-window navigation

**Comment by jungkees** [comment]
F2F: @jungkees should re-inspect the discussion here and the latest changes in HTML navigation.

**Comment by jungkees** [comment]
> That seems like an unfortunate setup since in general fetch doesn't need to know about this, only for navigation.

@annevk, to avoid that concern, we might want to invoke match the service worker registration twice: one for fetching the main resource in Handle Fetch and the other for setting a controller to the client's global. I outlined the simplified flow below:

```
// Window client matching
Navigate(request)

  Process a navigate fetch(request)

    Fetch(request)

      response = Handle Fetch(request)

        reg = Match Service Worker Registration(request.url)

        Dispatch the FetchEvent at reg's active worker

          // Use e.clientId, e.reservedClientId, e.targetClientId

      Process a navigate response(request, response)

        Load an HTML document

          Create a Document

          Initialize a new Document object

            Create a Window

            reg = Match Service Worker Registration(request.url or Document's URL?)
              // Are request.url and Document's URL always the same at this point?

            Set the Window object's active worker to reg's active worker

            Create an environment settings object
```

```
// Worker client matching
Run a worker

  Create a Dedicated/SharedWorkerGlobalScope

  Create a new environment settings object

  Fetch script

    response = Handle Fetch(request)

    reg = Match Service Worker Registration(request.url)

    Dispatch the FetchEvent at reg's active worker

      // Use e.clientId, e.reservedClientId, e.targetClientId

  (Before step 10) reg = Match Service Worker Registration(request.url)

  Set worker global's active worker to reg's active worker
```

There's still a possibility that the active worker of the registration can be changed from the first matching attempt and the second matching attempt. That is, in that occasion, the main resource and its subresources are served by two different service worker versions. @jakearchibald, I'd like to clarify if this would be okay.

**Comment by jungkees** [comment]
@jakearchibald, In addition, I'd like to revisit the use case of `e.reservedClientId` and `e.targetClientId` in the API perspective. It'll be great if you outline the usages of those attributes in some sample code.

To me, in my mental model, we need only two clients: the source and the target where the state of the target, `targetClient.created`, changes from false to true when the actual Global object is created.

In https://github.com/slightlyoff/ServiceWorker/issues/870#issuecomment-208487588, the `client.targetClientId` is set to the value of `client.reservedClientId` when the target is created. Or, are they two different values at the end?

**Comment by slightlyoff** [comment]
My suggestion is that these should be the same clientId at the end.

To add some urgency: It's bad that we can't tell if a fetch is targeted at a top-level document or an iframe anymore. The death of `isTopLevel` now appears premature. We had a question just today from a team that influences a notable % of overall web traffic if they could distinguish an iframe navigation for their PWA use-case. It's not possible and it makes their implementation nastier as a result. This is data we _have_. We should be reflecting what we know about the situation to the extent that we can do so safely.

**Comment by jungkees** [comment]
@slightlyoff, I think we can do something like this:

```js
self.onfetch = e => {
  if (e.request.destination === 'document') {
    // This is a navigation.
    clients.get(e.targetClientId).then(client => {
      if (client.ancestorOrigins.length === 0) {
        // And it's a top-level.
      }
    });
  }
};
```

**Comment by jakearchibald** [comment]
There's a lot of confusion around `targetId`, and reading back I've made some confusing statements too. Let me try and clarify:

* `clientId` - the client that initiated the request.
* `reservedClientId` - the client that the response may create. Undefined if the response may never result in a new client (subresources).
* `targetClientId` - the client that would become unloaded if the response creates a client. Undefined if the response is not replacing an existing client.

In https://github.com/w3c/ServiceWorker/issues/870#issuecomment-208498912 I suggested that `targetClientId` may sometimes be the same as `reservedClientId`. I think this is a mistake, and probably the cause of a lot of confusion in this thread - sorry! `targetClientId` will never be the same as `reservedClientId`.

# Subresrouce

`page1.html`:

```html
<script src="script.js"></script>
```

For the request for `script.js`:

* `clientId` - the client for `page1.html`
* `reservedClientId` - undefined as this is a subresource request
* `targetClientId` - undefined as this is a subresource request

# Simple navigation

`page1.html`

```html
<a href="page2.html">Hello</a>
```

When the link is clicked without modifiers, the request for `page2.html`:

* `clientId` - the client for `page1.html`
* `reservedClientId` - the client that will be used to display `page2.html`
* `targetClientId` - the same ID as `clientId`, as page2 replaces page1

# New window navigation

`page1.html`

```html
<a href="page2.html" target="_blank">Hello</a>
```

When the link is clicked without modifiers, the request for `page2.html`:

* `clientId` - the client for `page1.html`
* `reservedClientId` - the new client that will be used to display `page2.html`
* `targetClientId` - undefined, as page2 is not replacing anything

# Other window navigation

`page1.html`

```html
<iframe name="foo" src="page2.html">
<a href="page3.html" target="foo">Hello</a>
```

When the link is clicked without modifiers, the request for `page3.html`:

* `clientId` - the client for `page1.html`
* `reservedClientId` - the new client that will be used to display `page3.html`
* `targetClientId` - the client for `page2.html`

# Worker

`page1.html`

```html
<script>new Worker('script.js');</script>
```

The request for `script.js`:

* `clientId` - the client for `page1.html`
* `reservedClientId` - the new client that will be used to execute the worker
* `targetClientId` - undefined, nothing is being replaced

**Comment by jakearchibald** [comment]
@annevk is the above compatible with fetch's use of clients? That was my goal here

**Comment by jakearchibald** [comment]
If the naming of `targetClientId` is causing all this confusion, I'm up for renaming it. I used target because that's what's used on links. I don't really have a better suggestion though.

**Comment by jungkees** [comment]
I was misunderstanding the `targetClientId`. So, it's more like a client of the browsing context where the response's document loaded.

One question is what the use case for `targetClientId`  then?

As discussed on IRC, I understand `reservedClientId` can be used to link the two `FetchEvent`s: the main resource and its subresource.

**Comment by jungkees** [comment]
> One question is what the use case for targetClientId then?

Maybe the same reason? Like per-client cache management?

**Comment by jakearchibald** [comment]
The properties and methods of the reserved client are a little hacky, but here's my take:

```js
self.addEventListener('fetch', event => {
  if (event.reservedClientId) {
    clients.get(event.reservedClientId).then(c => console.log(c));
  }
});
```

* `url` - same as `event.request.url`
* `frameType` - do we have enough info to set the right value here?
* `id` - `event.reservedClientId`
* `postMessage()` - buffered until the client is created. Dropped if the client is never created.

Reserved window clients:

* `visibilityState` - the visibility state of the target window
* `focused` - the focused state of the target window
* `focus()` - rejects
* `navigate()` - rejects

**Comment by annevk** [comment]
@jakearchibald I think so, but it's not all defined in detail yet. Note that Fetch's current concept of target is a browsing context, not a client.

**Comment by jakearchibald** [comment]
@jungkees 

> Maybe the same reason? Like per-client cache management?

Yeah, if I know a client using cache "v1" is going away, I can get rid of that cache.

**Comment by jungkees** [comment]
> frameType - do we have enough info to set the right value here?

'frameType' will be replaced by `ancestorOrigins`. I think we can use target client's value here too as it may retain the same hierarchy of the ancestor browsing contexts. But that info is just valid under the assumption that the reserved client is to be **used**.

**Comment by jakearchibald** [comment]
A couple more details:

Any of the client IDs will be undefined if they point to a cross-origin client.

# about:blank

`page1.html`

```html
<iframe name="foo" src="about:blank">
<a href="page2.html" target="foo">Hello</a>
```

When the link is clicked without modifiers, the request for `page2.html`:

* `clientId` - the client for `page1.html`
* `reservedClientId` - the new client that will be used to display `page3.html`
* `targetClientId` - undefined, as `about:blank` does not have its own client, it uses the client for `page1.html`, which will not be replaced

**Comment by jungkees** [comment]
> @jakearchibald I think so, but it's not all defined in detail yet. Note that Fetch's current concept of target is a browsing context, not a client.

Yeah, but we use the same type (environment settings object) for **clients**. I think we should add an _id_ internal slot to environment settings object.

I think I need to define sort of a special environment settings object that has algorithms returning the values expected for reserved client during navigation (before the normal environment settings object is created.) And then when the normal environment settings object is created, the id of this new object is set to the id of the temporary environment settings object. From the script surface, `clients.get(id)` returns a `Client` object associated with either the temporary client or the created client depending on the state.

**Comment by jungkees** [comment]
We may need `client.used` in https://github.com/w3c/ServiceWorker/issues/870#issuecomment-245564424

**Comment by jungkees** [comment]
In https://github.com/w3c/ServiceWorker/issues/870#issuecomment-245559217's "New Window Navigation" case, `targetClientId` may be the client for initial `about:blank` instead of `undefined`. When `target="_blank"` is set a new browsing context is created loading the initial `about:blank` and then it navigates to the designated resource I guess.

**Comment by jakearchibald** [comment]
@jungkees 

> There's still a possibility that the active worker of the registration can be changed from the first matching attempt and the second matching attempt. That is, in that occasion, the main resource and its subresources are served by two different service worker versions. @jakearchibald, I'd like to clarify if this would be okay.

Yeah, that's fine. You opt in to dealing with that when you use `skipWaiting`.

> We may need client.used

Agreed!

> In #870 (comment)'s "New Window Navigation" case, targetClientId may be the client for initial about:blank instead of undefined

I was chatting to @annevk in #whatwg and he suggested that `about:blank` doesn't get a fetch client in this case, it uses the parent's.

**Comment by jungkees** [comment]
> I was chatting to @annevk in #whatwg and he suggested that about:blank doesn't get a fetch client in this case, it uses the parent's.

Ah.. right. So, I'll keep that in mind when I'm spec'ing this. But the value of the reserved client would not be undefined but the one represented by the environment settings object of that global I guess.

**Comment by jakearchibald** [comment]
Resolution:

* The ID sits on the env settings objects
* A reserved ID can be passed in which a new env settings object will use, otherwise it will generate its own

**Comment by jungkees** [comment]
Status note:

* Addressed the following (8b483b091e0f0bae6b698cf05d915c2029748ae0):
 - Use of request's reserved client (an environment or an environment settings object created during a navigation or a worker creation respectively in HTML) and request's target client id to expose `fetchEvent.reservedClientId/targetClientId` to fetch event handlers
 - Match a service worker registration and attach its active worker, if available, to the reserved client
* Changing the steps of Clients/Client's methods will be addressed in a separate patch.

**Comment by jungkees** [comment]
The issue has been addressed by adding/modifying the relevant concepts and steps in SW, HTML (navigate, run a worker, etc.), and Fetch. (See https://github.com/w3c/ServiceWorker/issues/870#issuecomment-256012207 for the related changes.)

Closing.
